### PR TITLE
Add CMake build and client support

### PR DESCRIPTION
**Pull request by friendlyanon**
What I need feedback on:

* What should be the package name of this project for CMake clients? Right now it's `async-mqtt5`, i.e. the package can be found with `find_package(async-mqtt5 REQUIRED)`
* What should the exported target name be? Right now it's `Async::MQTT5`.

I would've added GitHub Actions CI as well, but it seems the project has issues running its tests on Windows.  
Note that the project requires some additional defines on Windows to be set by the person configuring from the command line via `-DCMAKE_CXX_FLAGS=...` and that is good and must not be hardcoded into project code, because long term that is a fool's errand.

### CODE REVIEW

**Comment by ksimicevic** [comment]
Hello!

Thank you very much for your contribution! I'll try it out tomorrow. 

As for your questions, both names are fine right now.

**Comment by friendlyanon** [comment]
The command line I built the project with as a developer on Windows in a `-arch=amd64 -host_arch=amd64` `vcvars` enabled command prompt is:

```
cmake -B build -G Ninja -D async-mqtt5_DEVELOPER_MODE=1 -D CMAKE_BUILD_TYPE=Release -D "CMAKE_CXX_FLAGS=/sdl /guard:cf /utf-8 /diagnostics:caret /w14165 /w44242 /w44254 /w44263 /w34265 /w34287 /w44296 /w44365 /w44388 /w44464 /w14545 /w14546 /w14547 /w14549 /w14555 /w34619 /w34640 /w24826 /w14905 /w14906 /w14928 /w45038 /W4 /permissive- /volatile:iso /Zc:inline /Zc:preprocessor /Zc:enumTypes /Zc:lambda /Zc:__cplusplus /Zc:externConstexpr /Zc:throwingNew /EHsc /D_WIN32_WINNT=0x0A00 /DNTDDI_VERSION=0x0A000007 /DWIN32_LEAN_AND_MEAN=1 /DNOMINMAX=1 /bigobj" -D "CMAKE_EXE_LINKER_FLAGS=/machine:x64 /guard:cf" -D CMAKE_CXX_EXTENSIONS=0 -D CMAKE_CXX_STANDARD=20 -D CMAKE_CXX_STANDARD_REQUIRED=1 --toolchain "%VCPKG_ROOT%/scripts/buildsystems/vcpkg.cmake" -D VCPKG_MANIFEST_FEATURES=test
cmake --build build >build.log 2>&1
```

Redirecting the build output into a file is recommended, because the output from the warnings is plentiful. Produces a 10 MB output on my machine.

Don't mind the long configure command, it can be moved into a [preset](https://github.com/friendlyanon/cmake-init-header-only/blob/master/CMakePresets.json) at any time.

**Comment by friendlyanon** [comment]
Just a curiosity, I also tried to build with a couple other compilers on Windows as well:

* LLVM Clang 17.0.5: fails around the coroutine code
* ClangCL: fails around the coroutine code
* x86_64-13.2.0-release-win32-seh-ucrt-rt_v11-rev0: ICE around the coroutine code

**Comment by ksimicevic** [comment]
> Since this is a header-only library (for now), it should be documented that the user should find and link OpenSSL if they need the features. Putting this behind a flag for a header-only library is absolutely pointless and just makes project code more complex for no benefit.

The fact that OpenSSL is optional is documented in multiple places. However, I suppose it wouldn't hurt to note that in the CMakeLists.txt as well. Right now, a comment would be good enough. 

> That explains why they are giving problems. I think they ought to be buildable, otherwise you will not know if they are actually correct.

Of course! All the files are buildable except `network_connection.cpp`, which will be buildable if you comment out external customization points for TLS (for example, lines 139-156) as they are repeated for TLS and WebSocket over TLS examples. Creating multiple files for this case alone seemed like an overkill.

File: example/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+  add_executable("${name}" ${ARGN})
+  target_compile_features("${name}" PRIVATE cxx_std_20)
+  target_link_libraries("${name}" PRIVATE Async::MQTT5)
+endfunction()
+
+foreach(f callbacks cpp20_coroutines futures publisher receiver)
+  add_example("${f}" "${f}.cpp")
+endforeach()
+
+#[[
```
Comment (friendlyanon) [inline example/CMakeLists.txt:21]: I left this part in case you want to make this code compile. I can delete it if you want.

**Pull request merged**